### PR TITLE
First commit

### PR DESCRIPTION
**Pull request by dkess**
I've left some context about the meaning of this new repo in the README. But basically, this is part 1 of my solution for general user authentication in Kubernetes. Right now we don't have any particular use for authenticating individual users (as opposed to authenticating as just a staff member or ocfroot, for instance), but I have ideas for how this can help us down the line as we improve our developer experience with Kubernetes.

This PR isn't done. The `Dockerfile`, `Jenkinsfile`, and Apache config aren't here yet. They'll be added in a later commit. For now I'd like to hear people's thoughts on the code.

I know dealing with certificates is scary, but I am pretty confident that we can do this properly and correctly. Most of this code is taken straight from the Python `cryptography` library recipes.

### CODE REVIEW

**Comment by dkess** [comment]
A small addendum: this is inspired by how [MIT does auth](https://ca.mit.edu/ca/) in some parts of their infra. However I don't expect us to need a web UI for this, since I only expect us to use this from scripts.

**Comment by jvperrin** [comment]
You mention:

> Right now we don't have any particular use for authenticating individual users (as opposed to authenticating as just a staff member or ocfroot, for instance), but I have ideas for how this can help us down the line as we improve our developer experience with Kubernetes.

What is this for then? I suppose I'm still a bit confused on that front.

I'm also assuming there is a web server in front of this doing the Kerberos authentication work to convert their credentials and pass along the header needed?

**Comment by dkess** [comment]
Perhaps some background is needed.

Right now, the only way to get into the cluster is to SSH into one of the masters, become root, and use the admin keypair. The statement in parentheses is drawing attention to an alternate authentication scheme which I chose not to use, despite it being easier to implement. Such a scheme would involve a shared credential of sorts which all staffers could use (perhaps a reverse proxy webserver accessible only on supernova), and as such wouldn't authenticate users as individuals, only as members of an "ocfstaff" group.

The system in this PR could be used to authenticate actual users, which isn't useful now (everyone would have the same permissions anyways) but could be useful down the line if we want to do more exotic things with the cluster. For instance, I envision letting staffers test stuff out in their own namespace

To your second question: yes, the plan is to add Apache behind this to handle Kerberos, similar to how we used to do it with RT.

File: app.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+private_key = load_pem_private_key(
+    data=open(PRIVATE_KEY_FILE, 'rb').read(),
+    password=None,
+    backend=default_backend(),
+)
+
+
+def user_groups(username):
+    """Returns all the groups a user is in"""
+    return [g.gr_name for g in grp.getgrall() if username in g.gr_mem]
```
Comment (kpengboy) [inline app.py:30]: This wouldn't get the user's primary group, would it?

File: app.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+app = Flask(__name__)
+
+
+@app.route('/cert', methods=['POST'])
+def cert():
+    # TODO: figure out what this header is actually called
+    username = request.headers['AUTH_USER']
+
+    req_json = request.get_json()
+    print(req_json)
```
Comment (kpengboy) [inline app.py:42]: Don't forget to remove before merging

File: app.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+    builder = builder.issuer_name(x509.Name([
+        x509.NameAttribute(
+            NameOID.COMMON_NAME,
+            'krb-kubernetes.ocf.berkeley.edu',
+        ),
+    ]))
+
+    # Set the cert's expiration
+    builder = builder.not_valid_before(datetime.datetime.today())
```
Comment (kpengboy) [inline app.py:68]: Maybe subtract a minute in case of clock drift?

File: app.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+@app.route('/cert', methods=['POST'])
+def cert():
+    # TODO: figure out what this header is actually called
+    username = request.headers['AUTH_USER']
+
+    req_json = request.get_json()
+    print(req_json)
+    pubkey = load_pem_public_key(
+        req_json['pubkey'].encode(), backend=default_backend(),
```
Comment (kpengboy) [inline app.py:44]: I would specify an encoding to be extra safe

File: app.py [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+    )
+
+    certificate = builder.sign(
+        private_key=private_key,
+        algorithm=hashes.SHA256(),
+        backend=default_backend(),
+    )
+
+    return jsonify(
+        cert=certificate.public_bytes(serialization.Encoding.PEM).decode(),
```
Comment (kpengboy) [inline app.py:89]: I would specify an encoding to be extra safe

File: certsign.wsgi [hunk 6]
```diff
@@ -0,0 +1 @@
+from app import app as application  # NOQA
```
Comment (kpengboy) [inline certsign.wsgi:1]: Can a more specific disable be used here?

**Comment by dkess** [comment]
I am going to close this since I have a better idea for how to implement this. The code will be similar but will not be accessible as an HTTP service, instead it'll be a script inside Puppet. I'll submit a new PR and also try to address the feedback there.

**Pull request closed without merging**